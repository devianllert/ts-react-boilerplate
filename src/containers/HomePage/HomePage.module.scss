@import '../../design/design-system';

.header {
  display: flex;
  flex-direction: column;

  justify-content: center;
  align-items: center;

  min-height: 360px;
}

.title {
  margin-top: 0;

  font-size: $font-size-h2;
  line-height: $line-height-h2;

  text-align: center;
}

.subtitle {
  margin-top: 0;

  font-size: $font-size;
  line-height: $line-height;

  text-align: center;
}