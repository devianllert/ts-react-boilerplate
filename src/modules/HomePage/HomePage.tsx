import React, { ReactElement } from 'react';
import { useTranslation } from 'react-i18next';
import { Helmet } from 'react-helmet';

import Typography from '../../components/Typography';
import Container from '../../components/Container';

const HomePage = (): ReactElement => {
  const { t } = useTranslation();

  return (
    <>
      <Helmet>
        <title>{t('HOME_TITLE')}</title>
      </Helmet>

      <Container>
        <Typography variant="h1" align="center" gutterBottom>Home</Typography>
        <Typography variant="h2" align="center" paragraph>React typescript boilerplate</Typography>
      </Container>
    </>
  );
};

export default HomePage;
