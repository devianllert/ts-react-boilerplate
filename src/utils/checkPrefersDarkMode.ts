/**
 * Check browser or OS preference for dark mode
 */

const checkPrefersDarkMode = (): boolean => {
  const preferDarkQuery = '(prefers-color-scheme: dark)';

  const mediaQuery = matchMedia(preferDarkQuery);

  return mediaQuery.media === preferDarkQuery;
};

export default checkPrefersDarkMode;
