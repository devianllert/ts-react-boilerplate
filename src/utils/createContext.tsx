import React, {
  createContext,
  useState,
  useContext,
  Provider,
  Dispatch,
  SetStateAction,
  PropsWithChildren,
  ReactNode,
  Context,
} from 'react';

type UpdateType<A> = Dispatch<SetStateAction<A>>;

/**
 * A helper to create a Context and Provider with no upfront default value, and
 * without having to check for undefined all the time.
 */
export const createCtx = <A extends {} | null>(): readonly [() => A, Provider<A | undefined>] => {
  const ctx = createContext<A | undefined>(undefined);

  const useCtx = (): A => {
    const c = useContext(ctx);

    if (c === undefined) {
      throw new Error('useCtx must be inside a Provider with a value');
    }

    return c;
  };

  return [useCtx, ctx.Provider] as const;
};

export const createCtxWithUpdate = <A extends any>(
  defaultValue: A,
): readonly [
  Context<{ state: A; update: Dispatch<SetStateAction<A>> }>,
  (props: { children?: ReactNode }) => ReactNode,
] => {
  const defaultUpdate: UpdateType<typeof defaultValue> = () => defaultValue;

  const ctx = createContext({
    state: defaultValue,
    update: defaultUpdate,
  });

  const ContextProvider = (props: PropsWithChildren<{}>): ReactNode => {
    const [state, update] = useState(defaultValue);
    // eslint-disable-next-line react/jsx-props-no-spreading
    return <ctx.Provider value={{ state, update }} {...props} />;
  };

  return [ctx, ContextProvider] as const;
};
