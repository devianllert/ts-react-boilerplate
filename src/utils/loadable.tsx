/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable react/jsx-props-no-spreading */

import React, {
  lazy,
  Suspense,
  ReactElement,
  ComponentType,
} from 'react';

export type ImportComponent = Promise<{ default: any }>;

const loadable = (importFunc: () => ImportComponent, { fallback = null }): ComponentType => {
  const LazyComponent: ComponentType = lazy(importFunc);

  return (props: object): ReactElement => (
    <Suspense fallback={fallback}>
      <LazyComponent {...props} />
    </Suspense>
  );
};

export default loadable;
