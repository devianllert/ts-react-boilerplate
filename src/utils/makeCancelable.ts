/* eslint-disable prefer-promise-reject-errors */
/* https://reactjs.org/blog/2015/12/16/ismounted-antipattern.html */
export interface CancellablePromise<T> {
  promise: Promise<T>;
  cancel(): void;
}

const makeCancelable = <T>(promise: Promise<T>): CancellablePromise<T> => {
  let hasCanceled = false;

  const wrappedPromise = new Promise<T>((resolve, reject) => {
    promise.then(
      (val) => (hasCanceled ? reject({ isCanceled: true }) : resolve(val)),
      (error) => (hasCanceled ? reject({ isCanceled: true }) : reject(error)),
    );
  });

  return {
    promise: wrappedPromise,
    cancel(): void {
      hasCanceled = true;
    },
  };
};

export default makeCancelable;
