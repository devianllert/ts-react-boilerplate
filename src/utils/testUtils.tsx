/* eslint-disable import/prefer-default-export */
import React, { ReactElement } from 'react';
import { Router } from 'react-router-dom';
import { createMemoryHistory, MemoryHistory } from 'history';
import { Provider } from 'react-redux';
import { i18n as i18Next } from 'i18next';
import { I18nextProvider } from 'react-i18next';
import { render, RenderResult } from '@testing-library/react'; // eslint-disable-line

import configureStore, { EnhancedStore } from '../core/configureStore';
import i18nInstance from '../core/i18n';

export interface RenderProvidersOptions {
  initialState?: {};
  route?: string;
  routerHistory?: MemoryHistory;
  store?: EnhancedStore;
  i18n?: i18Next;
}

export interface RenderProvidersResult extends RenderResult {
  routerHistory: MemoryHistory;
  store: EnhancedStore;
  i18n: i18Next;
}

export type RenderProviders = (ui: ReactElement, options?: RenderProvidersOptions) => RenderProvidersResult;

const renderWithProviders: RenderProviders = (
  ui,
  {
    initialState = {},
    route = '/',
    routerHistory = createMemoryHistory({ initialEntries: [route] }),
    store = configureStore(initialState),
    i18n = i18nInstance,
  }: RenderProvidersOptions = {},
): RenderProvidersResult => ({
  ...render(
    <Provider store={store}>
      <I18nextProvider i18n={i18n}>
        <Router history={routerHistory}>{ui}</Router>
      </I18nextProvider>
    </Provider>,
  ),
  // adding `history` to the returned utilities to allow us
  // to reference it in our tests (just try to avoid using
  // this to test implementation details).
  store,
  routerHistory,
  i18n,
});

export {
  renderWithProviders,
};
