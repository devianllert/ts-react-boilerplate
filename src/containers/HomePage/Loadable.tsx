import loadable, { ImportComponent } from '../../utils/loadable';

const LoadableHomePage = loadable((): ImportComponent => import(/* webpackChunkName: "homePage" */ './HomePage'));

export default LoadableHomePage;
