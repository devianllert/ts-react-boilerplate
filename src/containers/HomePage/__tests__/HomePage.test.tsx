import React from 'react';
import { cleanup } from '@testing-library/react';

import { renderWithRouter } from '../../../utils/testUtils';

import HomePage from '../HomePage';

import '@testing-library/jest-dom/extend-expect';

describe('<HomePage />', (): void => {
  afterEach(cleanup);

  it('should render and match the snapshot', (): void => {
    const { container } = renderWithRouter(<HomePage />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render with initial state', (): void => {
    const { getByText } = renderWithRouter(<HomePage />);

    expect(getByText(/home.title/)).toBeDefined();
  });
});
