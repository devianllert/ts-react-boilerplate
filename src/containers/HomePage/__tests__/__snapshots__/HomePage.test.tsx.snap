// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomePage /> should render and match the snapshot 1`] = `
<div
  class="header"
>
  <h1
    class="title"
  >
    home.title
  </h1>
  <p
    class="subtitle"
  >
    home.subtitle
  </p>
</div>
`;
